File: android/app/src/main/java/es/manabe/yomiyasu/core/models/StatsModels.kt
package es.manabe.yomiyasu.core.models

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

@Serializable
data class UserStats(
    val totalMangaBooks: Int = 0,
    val totalNovelaBooks: Int = 0,
    val totalPagesRead: Int = 0,
    val totalCharacters: Int = 0,
    val totalMangaSeries: Int = 0,
    val totalNovelaSeries: Int = 0,
    val totalTimeRead: Double = 0.0,
)

@Serializable
data class MonthlyGraphEntry(
    @SerialName("_id") val monthId: MonthId,
    val totalCharacters: Int = 0,
    val totalTime: Int = 0,
    val meanReadSpeed: Double = 0.0,
    val totalHours: Double = 0.0,
) {
    val id: String get() = "${monthId.year}-${monthId.month}"

    @Serializable
    data class MonthId(
        val year: Int = 0,
        val month: Int = 0,
    )
}

@Serializable
data class MonthlyGraphs(
    val manga: List<MonthlyGraphEntry> = emptyList(),
    val novela: List<MonthlyGraphEntry> = emptyList(),
)

@Serializable
data class StreakDay(
    val dayOfMonth: Int,
    val count: Int = 0,
)

@Serializable
data class ProgressRecord(
    @SerialName("_id") val id: String,
    val book: String? = null,
    val serie: String? = null,
    val startDate: String? = null,
    val lastUpdateDate: String? = null,
    val endDate: String? = null,
    val time: Int? = null,
    val currentPage: Int? = null,
    val status: ProgressStatus? = null,
    val paused: Boolean? = null,
    val characters: Int? = null,
    val variant: Variant? = null,
    val day: Int? = null,
    val month: Int? = null,
    val year: Int? = null,
    val bookInfo: Book? = null,
    val serieInfo: Serie? = null,
    val meanReadSpeed: Double? = null,
) {
    val bookName: String get() = bookInfo?.visibleName ?: "—"
    val serieName: String get() = serieInfo?.visibleName ?: ""
    val resolvedVariant: Variant get() = variant ?: bookInfo?.variant ?: Variant.Manga

    val logLine: String
        get() {
            var text = when (resolvedVariant) {
                Variant.Manga -> ".log manga ${currentPage ?: 1} $bookName"
                Variant.Novela -> ".log lectura ${characters ?: 0} $bookName"
            }
            val minutes = time
            if (minutes != null && minutes > 59) {
                text += ";${minutes / 60}"
            }
            if (resolvedVariant == Variant.Manga) {
                val chars = characters
                if (chars != null && chars > 0) {
                    text += "&$chars"
                }
            }
            return text
        }
}

@Serializable
data class CreateReviewRequest(
    val serie: String,
    val userLevel: String,
    val difficulty: Int,
    val valoration: Int,
    val comment: String,
)

enum class ReviewLevel(val rawValue: String) {
    Beginner("Principiante"),
    N5("N5"),
    N4("N4"),
    N3("N3"),
    N2("N2"),
    N1("N1"),
    N1Plus("N1+"),
}

File: ios/Yomiyasu/Core/Models/StatsModels.swift
import Foundation

struct UserStats: Decodable, Sendable {
    let totalMangaBooks: Int
    let totalNovelaBooks: Int
    let totalPagesRead: Int
    let totalCharacters: Int
    let totalMangaSeries: Int
    let totalNovelaSeries: Int
    let totalTimeRead: Double
}

struct MonthlyGraphEntry: Decodable, Sendable, Identifiable {
    struct MonthID: Decodable, Sendable, Hashable {
        let year: Int
        let month: Int
    }

    let monthID: MonthID
    let totalCharacters: Int
    let totalTime: Int
    let meanReadSpeed: Double
    let totalHours: Double

    var id: String { "\(monthID.year)-\(monthID.month)" }

    private enum CodingKeys: String, CodingKey {
        case monthID = "_id"
        case totalCharacters
        case totalTime
        case meanReadSpeed
        case totalHours
    }
}

struct MonthlyGraphs: Decodable, Sendable {
    let manga: [MonthlyGraphEntry]
    let novela: [MonthlyGraphEntry]
}

struct StreakDay: Decodable, Sendable, Identifiable {
    let dayOfMonth: Int
    let count: Int

    var id: Int { dayOfMonth }
}

struct ProgressRecord: Decodable, Sendable, Identifiable {
    let id: String
    let book: String?
    let serie: String?
    let startDate: Date?
    let lastUpdateDate: Date?
    let endDate: Date?
    let time: Int?
    let currentPage: Int?
    let status: ProgressStatus?
    let paused: Bool?
    let characters: Int?
    let variant: Variant?
    let day: Int?
    let month: Int?
    let year: Int?
    let bookInfo: Book?
    let serieInfo: Serie?
    let meanReadSpeed: Double?

    private enum CodingKeys: String, CodingKey {
        case id = "_id"
        case book
        case serie
        case startDate
        case lastUpdateDate
        case endDate
        case time
        case currentPage
        case status
        case paused
        case characters
        case variant
        case day
        case month
        case year
        case bookInfo
        case serieInfo
        case meanReadSpeed
    }
}

extension ProgressRecord {
    var bookName: String { bookInfo?.visibleName ?? "—" }
    var serieName: String { serieInfo?.visibleName ?? "" }
    var resolvedVariant: Variant { variant ?? bookInfo?.variant ?? .manga }

    var logLine: String {
        var text: String

        switch resolvedVariant {
        case .manga:
            text = ".log manga \(currentPage ?? 1) \(bookName)"
        case .novela:
            text = ".log lectura \(characters ?? 0) \(bookName)"
        }

        if let time, time > 59 {
            text += ";\(time / 60)"
        }

        if resolvedVariant == .manga, let characters, characters > 0 {
            text += "&\(characters)"
        }

        return text
    }
}

struct CreateReviewRequest: Encodable, Sendable {
    let serie: String
    let userLevel: String
    let difficulty: Int
    let valoration: Int
    let comment: String
}

enum ReviewLevel: String, CaseIterable, Identifiable {
    case beginner = "Principiante"
    case n5 = "N5"
    case n4 = "N4"
    case n3 = "N3"
    case n2 = "N2"
    case n1 = "N1"
    case n1Plus = "N1+"

    var id: String { rawValue }
}
